### PR TITLE
feat: `@listSuffix omit/include` smart tag

### PR DESCRIPTION
**Pull request by keithlayne**
## Description

I've been using this as a patch for several months, and kept forgetting
to make a PR or open an issue about it. The use case is for when you are
*not* using `pgSimpleCollections = "only"` and `pgOmitListSuffix`, but
want to essentially apply those selectively.

I don't exactly remember the motivating case, or if it covers much
behavior.  Obviously needs tests, but throwing it out there for comment.

## Performance impact

Looks like it has a small perf impact, it adds a couple of ternary expressions.

## Security impact

unknown

## Checklist

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [x] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [x] If this is a breaking change I've explained why.

### CODE REVIEW

**Review by benjie — CHANGES_REQUESTED** [review]
This is neat; thanks for the PR. Please could you:

1. do the opposite for the connection equivalents (so that a naming collision does not occur); i.e. if there's `omitListSuffix` on something but it has both simple collections and relay connections then the relay connection should have `Connection` appended
2. add details to the README of this tag

Thanks!

DISCLAIMER: I'm currently recovering from sedation.

**Comment by keithlayne** [comment]
Ha!  So you only like it due to your questionable mental status?  :) 

Anyway, just wanted to get your input before moving forward - I think I need to take a more general approach here.  It should consider the global settings and then per-entity tags for both simple and relay naming.  If I can do that, I think this particular configuration will be covered pretty completely for user tweaking.

Of course, tests and docs might be most of the work.  Fun!

**Comment by benjie** [comment]
I only vaguely recall writing the above :laughing: You're right, I don't think there's much actual code you need to write, but there's the tests and the README.

**Comment by keithlayne** [comment]
Got to thinking about this, and part of me thinks the whole thing can be 
simplified a lot with...a breaking change.  Here are some notes I made while I 
was thinking about it:

If the effective value of `simpleCollections` for any particular field is
`@simpleCollections` if present, defaulting to the `--simpleCollections` config,
and `omitListSuffix` is treated similarly with a new `@omitListSuffix` smart
tag, here are the possible combinations (`X` means it gets a suffix):

(This is not verified to be how it actually works, but is intuitively what I
expect)

| simpleCollections | omitListSuffix | List | Connection |
| :---------------: | :------------: | :--: | :--------: |
|       omit        |     false      | N/A  |            |
|       omit        |      true      | N/A  |            |
|       only        |     false      |  X   |    N/A     |
|       only        |      true      |      |    N/A     |
|       both        |     false      |  X   |            |
|       both        |      true      |      |     X      |

Thoughts:

- When `simpleCollections` is `omit` or `only` for a particular field, we should
  need no tags - the collection should not have a suffix
- This should only be relevant when (for a particular field),
  `simpleCollections` is `both`
- Possibly instead of `pgOmitListSuffix`, there should be a
  `pgOmitCollectionSuffix` / `@omitCollectionSuffix` setting of type
  `'connection' | 'list'` with the default being unset. It would only be
  relevant when both collections types are generated.
- I think this gives full flexibility, but it _is_ a breaking change. What am I
  missing or forgetting?
- Do some people _want_ the suffix when they are using only simple collections?
  In order to provide an upgrade path while preserving present the present
  schema, seems like there should be an option for that.
- Should the `@name` tag be considered in some or all of these cases
  after/instead of the inflector name, versus the current behavior of acting as
  a base for the suffix?

One option is to not have an `omit` tag, but more of an `add` tag. To preserve
current behavior, it could be `'list' | 'connection'` and default to `'list'`. In this
case, if both collections are gererated, you can specify `'connection'` for that
one to get the suffix instead of the default.

**Comment by benjie** [comment]
> I think this gives full flexibility, but it is a breaking change. What am I missing or forgetting?

I'm very big on names not changing depending on other things in the schema. For example; we don't check to see if there are other foreign keys before naming a relation - if we did then later adding another foreign key might change the name of the original reference, which is 100% undesired. Similarly if you have a simple collection called `foos` and then you turn on the relay connection, `foos` should not be renamed to `foosList` and the connection added as `foos` - that would be very confusing (and breaking).

As such; I don't like chosing the name of things based on other settings. For me:

- `omitListSuffix=false` means that connections have no suffix and lists have `List` suffix
- `omitListSuffix=true` means that connections get `Connection` suffix and lists have no suffix

That's the (simple) rule for that configuration option; and it means your schema is consistent, and even as you turn on different connection/collection settings no existing fields will be renamed.

The `simpleCollections only/omit/both` setting only controls which fields show up, NOT what they're named.

As such these two settings are completely independent, and the user does not have to worry themselves about the "truth table" of how they combine. What you lay out may make intuitive sense assuming that you never change any of these settings once you first set them, but they become a maintenance burden because you can no longer just turn on "simpleCollections both" and expect everything to just work, now you have to also remember to invert "omitListSuffix". And imagine if there was later a third tag that interacted too ─ the complexity would very quickly escalate.

---

In conclusion: your original proposal of using an `omitListSuffix` tag on a specific resource to override the global setting in that one place is the desired approach.

**Comment by keithlayne** [comment]
Thanks for sharing your thoughts.  Sounds good.  At the end of the day, I just want to be able to control the names.  IIRC the only thing that's missing from my patch for my use case is top-level `simpleCollections=only` queries.  I just don't want to miss anything.

**Comment by benjie** [comment]
Where you've overridden `computedColumnList` you should also override `computedColumn` to do the inverse; etc.

**Comment by keithlayne** [comment]
@Benjie can you take a quick look when you get a chance and let me know if this approach is viable?  Don't know if the function call overhead is bad here.

**Comment by keithlayne** [comment]
I was confused for a minute because tests silently fail to run in node 14. The dev dep on postgraphile was `4.6.0` and locally I have set it to `4.9.0` but I'm not sure what the right thing is to do there. Not going to commit that unless you tell me to, just a point of confusion.

**Comment by benjie** [comment]
Tests: remove the lockfile and run `yarn` again. Node 14 broke the `pg` module. Also upgrade to pg@8 I think.

File: index.js [hunk 1]
```diff
@@ -48,10 +48,14 @@ function PgSimplifyInflectorPlugin(
...
-  const listSuffix = pgOmitListSuffix ? "" : "-list";
-  const ListSuffix = pgOmitListSuffix ? "" : "List";
+  const connectionSuffix = (omitListSuffixTag) =>
+    pgOmitListSuffix || omitListSuffixTag ? "-connection" : "";
+  const ConnectionSuffix = (omitListSuffixTag) =>
+    pgOmitListSuffix || omitListSuffixTag ? "Connection" : "";
+  const listSuffix = (omitListSuffixTag) =>
+    gOmitListSuffix || omitListSuffixTag ? "" : "-list";
+  const ListSuffix = (omitListSuffixTag) =>
+    pgOmitListSuffix || omitListSuffixTag ? "" : "List";
```
Comment (benjie) [inline index.js]: If `omitListSuffix` is set to `0` we should make it unset `pgOmitListSuffix`. So `@omitListSuffix` -> true; `@omitListSuffix 0` -> false. Then if the tag is set, it should rule; otherwise pgOmitListSuffix is used as the fallback.

```suggestion
  function o(tag) { return tag != null ? tag !== "0" : !!pgOmitListSuffix; }
  const connectionSuffix = (omitListSuffixTag) =>
    o(omitListSuffixTag) ? "-connection" : "";
  const ConnectionSuffix = (omitListSuffixTag) =>
    o(omitListSuffixTag) ? "Connection" : "";
  const listSuffix = (omitListSuffixTag) =>
    o(omitListSuffixTag) ? "" : "-list";
  const ListSuffix = (omitListSuffixTag) =>
    o(omitListSuffixTag) ? "" : "List";
```

Comment (benjie) [reply index.js]: Alternatively we could use `@omitListSuffix` / `@includeListSuffix` tags, but that's kinda awkward.

Actually, maybe `@listSuffix include/omit` would be better...?


```suggestion
  function o(tag) { return tag != null ? tag === "omit" : !!pgOmitListSuffix; }
  const connectionSuffix = (tag) =>
    o(tag) ? "-connection" : "";
  const ConnectionSuffix = (tag) =>
    o(tag) ? "Connection" : "";
  const listSuffix = (tag) =>
    o(tag) ? "" : "-list";
  const ListSuffix = (tag) =>
    o(tag) ? "" : "List";
```

**Review by benjie — COMMENTED** [review]
I like the approach :+1:

File: index.js (hunk 1 above)
Comment (keithlayne) [reply index.js]: My thinking would be that it would be strictly additive - there's always a workaround to set to to false globally then enable it everywhere you want it.  Incidentally, I leave it off globally, otherwise all my connections (the vast majority of things) would have `Connection` suffixes.

Dumb question: with the restrictions on valid tag inputs, I assume you mean a string `0` there?  I would probably prefer some words there, and would also prefer to use a single tag if feasible.  But I agree the UX is better if you can override the base setting selectively in both directions.

**Comment by keithlayne** [comment]
Regarding the deps, this lib has a dev dependency on postgraphile `=4.6.0`, and no direct dependency on `pg`.  4.6.0 has a `<8` dependency on `pg`, so I just bumped to 4.9.0 locally for now.  I don't know that its really an issue, but I defer to you.

File: index.js (hunk 1 above)
Comment (keithlayne) [reply index.js]: I added some tests around what I had already written just to check my assumptions.

I will rework them with the latter set of changes - I wasn't sure if you'd made up your mind.  One argument against multiple tags: you'd have to handle when they're both set with an error or warning or silently picking a winner.  I don't think that's a very realistic concern, but I'd rather not have to worry about it.

I was looking at simplifying things for some cases when you currently need `@foreignSimpleFieldName`, but I found out that apparently this plugin runs too late to deconflict name clashes in that case.  Generally when that tag is in play then the suffix tag is effectively a no-op.

Comment (benjie) [reply index.js]: >  I assume you mean a string `0` there?

Yep. Smart tags don't have numbers.

> you'd have to handle when they're both set with an error or warning or silently picking a winner

Yeah, I came to this conclusion too, which is why I suggested the `@listSuffix` singular tag.

File: index.js [hunk 2]
```diff
@@ -48,10 +48,21 @@ function PgSimplifyInflectorPlugin(
     );
   }
 
-  const connectionSuffix = pgOmitListSuffix ? "-connection" : "";
-  const ConnectionSuffix = pgOmitListSuffix ? "Connection" : "";
-  const listSuffix = pgOmitListSuffix ? "" : "-list";
-  const ListSuffix = pgOmitListSuffix ? "" : "List";
+  function omitListSuffix(tag) {
+    return tag != null ? tag === "omit" : !!pgOmitListSuffix;
```
Comment (benjie) [inline index.js]: Should we thrown if the value is unknown? We'd expect `=== "omit"` or `=== "include"`; all other values should throw I think.

**Review by benjie — CHANGES_REQUESTED** [review]
This is feeling right; and it's a non-breaking change. Excellent work! How are you feeling about it?

Note to self: you've not yet reviewed the snapshots.

File: index.js (hunk 2 above)
Comment (keithlayne) [reply index.js]: I started updating the readme but didn't finish over the weekend. BTW I usually prefer to use arrow functions where I can (my opinion on function hoisting fluctuates) but changed them because it seemed to fit the style used in the file better.

Would you also want to reject `true` as a valid value? That seems reasonable to me but just wanted to make sure.

Since you brought it up I guess throwing is the way to handle this kind of error? I know you have some non-fatal error handling in there, but I don't know how that works.

Comment (benjie) [reply index.js]: Lets treat `@listSuffix` (i.e. `true`) as if it's `@listSuffix include`. I think that's fairly intuitive. Or throw; I'm happy either way.

**Comment by keithlayne** [comment]
Oops, didn't see your comment about `true` so I'm throwing if that's the value. I can fix if you want, otherwise I think this is ready for you to look at.

I tried to slim down the test output; in the latest the diffs are significantly smaller.

File: README.md [hunk 3]
```diff
@@ -185,3 +187,26 @@ suffix" ("List" by default, but "" if `pgOmitListSuffix` is set), but if you
...
+
+- foreign key constraints
+
+### `@listSuffix`
+
+`@listSuffix` allows you to override the default naming on a per-entity basis,
+overriding `pgOmitListSuffix`. For example, with `pgOmitListSuffix: true`, you
+can apply `@listSuffix include` to have the list suffix appended to the simple
+collection generated for that table. When `pgOmitListSuffix` is not `true`, then
+you can use `@listSuffix omit` to selectively omit the list suffix on entities.
```
Comment (benjie) [inline README.md]:
```suggestion
`@listSuffix` allows you to override the default naming on a per-entity basis,
overriding `pgOmitListSuffix`. For example, with `pgOmitListSuffix: true`, you
can apply `@listSuffix include` to have the `-List` suffix appended to the simple
collection generated for that table, and remove the `-Connection` suffix from the
Relay connection. When `pgOmitListSuffix` is not `true`, you can use
`@listSuffix omit` to selectively omit the `-List` suffix on simple collections
and append `-Connection` to the Relay connection instead.
```

File: README.md [hunk 4]
```diff
@@ -185,3 +187,26 @@ suffix" ("List" by default, but "" if `pgOmitListSuffix` is set), but if you
...
+
+If `@listSuffix` is set, the only valid values are `"omit"` and `"include"`. Any
+other value will cause an error.
```
Comment (benjie) [inline README.md:206]:
```suggestion
other value will cause an error.

|                   | @listSuffix omit    | @listSuffix include |
| ----------------: | :------------------ | :------------------ |
|  Relay Connection | companiesConnection | companies           |
| Simple Collection | companies           | companiesList       |
```

**Review by benjie — CHANGES_REQUESTED** [review]
Looks great; thanks @keithlayne! :raised_hands:

**Review by benjie — APPROVED** [review]
